From c66f8a5db4ae60752d8aad4b1850568445ceba9c Mon Sep 17 00:00:00 2001
From: fafryd <dz1125.bug.tracker@gmail.com>
Date: Thu, 5 Sep 2019 22:15:23 +0200
Subject: [PATCH] meson: bump gl >= 1.2

---
 glx/meson.build | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/glx/meson.build b/glx/meson.build
index 7f9e549f0..1140522f6 100644
--- a/glx/meson.build
+++ b/glx/meson.build
@@ -39,7 +39,7 @@ if build_glx
             common_dep,
             dl_dep,
             dependency('glproto', version: '>= 1.4.17'),
-            dependency('gl', version: '>= 9.2.0'),
+            dependency('gl', version: '>= 1.2'),
         ],
         c_args: [
             glx_align64,
@@ -74,7 +74,7 @@ if build_glx
             common_dep,
             dl_dep,
             dependency('glproto', version: '>= 1.4.17'),
-            dependency('gl', version: '>= 9.2.0'),
+            dependency('gl', version: '>= 1.2'),
         ],
     )
 
-- 
2.23.0

